extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java

```diff
@@ -45,34 +46,173 @@
+                    }
+                }
+            }
+        }
+
+        public Object getValueAtOffset(int idxAtPartition,
+                                       int offset,
+                                       WindowFrameState currentFrame,
+                                       List<? extends CollectExpression<Row, ?>> expressions,
+                                       boolean ignoreNulls,
+                                       Input[] args) {
+            if (ignoreNulls == false) {
+                var targetIndex = getTargetIndex(idxAtPartition, offset);
+                return getValueAtTargetIndex(targetIndex, currentFrame, expressions, args);
+            } else {
+                return getValueAtOffsetIgnoringNulls(idxAtPartition, offset, currentFrame, expressions, args);
+            }
+        }
+
+        static Object getValueAtTargetIndex(Integer targetIndex,
+                                            WindowFrameState currentFrame,
+                                            List<? extends CollectExpression<Row, ?>> expressions,
+                                            Input[] args) {
+            if (targetIndex == null) {
+                throw new IndexOutOfBoundsException();
+            }
+            Object[] rowCells = currentFrame.getRowInPartitionAtIndexOrNull(targetIndex);
+            if (rowCells == null) {
+                throw new IndexOutOfBoundsException();
+            }
+            var row = new RowN(rowCells);
+            for (CollectExpression<Row, ?> expression : expressions) {
+                expression.setNextRow(row);
+            }
+            return args[0].value();
+        }
+    }
+
+    private final Supplier<OffsetDirection> FORWARD = () -> new OffsetDirection() {
+
+        @Override
+        int getTargetIndex(int idxInPartition, int offset) {
+            return idxInPartition + offset;
+        }
+
+        @Override
+        Object getValueAtOffsetIgnoringNulls(int idxInPartition,
+                                             int offset,
+                                             WindowFrameState currentFrame,
+                                             List<? extends CollectExpression<Row, ?>> expressions,
+                                             Input[] args) {
+            if (cachedNonNullIndex == null) {
+                cachedNonNullIndex = findNonNullOffsetFromCurrentIndex(idxInPartition, offset, currentFrame, expressions, args);
+            } else {
+                var curValue = getValueAtTargetIndex(idxInPartition, currentFrame, expressions, args);
+                if (curValue != null) {
+                    moveCacheToNextNonNull(currentFrame, expressions, args);
+                }
+            }
+            return getValueAtTargetIndex(cachedNonNullIndex, currentFrame, expressions, args);
+        }
+    };
+
+    private final Supplier<OffsetDirection> BACKWARD = () -> new OffsetDirection() {
```

Comment:
What's the reason for wrapping this in a `Supplier` ? It doesn't contain any state?